File: src/main/java/service/Service.java
package service;

import service.exception.ServiceException;

import java.util.List;

public interface Service<T> {
    List<T> listarTodos();

    T salvar(T entity) throws ServiceException;

    T atualizar(T entity) throws ServiceException;

    T encontraPeloId(T entity) throws ServiceException;
}

File: src/main/java/service/Service.kt
package service

import service.exception.ServiceException

interface Service<T> {
    fun listarTodos(): List<T>

    @Throws(ServiceException::class)
    fun salvar(entity: T): T

    @Throws(ServiceException::class)
    fun atualizar(entity: T): T

    @Throws(ServiceException::class)
    fun encontraPeloId(entity: T): T
}